Fix part of #4057 Make PlayerPositionService.js 100% test coverage

ujjwalgulecha opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR fixes part of #4057 PlayerPositionService.js making the coverage 100%

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

<img width="1275" alt="screen shot 2019-02-12 at 11 10 46 pm" src="https://user-images.githubusercontent.com/5632068/52693414-86945000-2f1b-11e9-828f-202bf273b345.png">

ujjwalgulecha committed:
Fix #4057 Make PlayerPositionService.js 100% test coverage

ujjwalgulecha commented:
@ankita240796 @vinitamurthi if you could please review or assign a reviewer. Thanks!

ankita240796 commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
```
Change year to 2019.

ujjwalgulecha replied:
changed

apb7 commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
```
`2015` --> `2019`

ujjwalgulecha replied:
changed

ankita240796 reviewed (changes requested):
Hi @ujjwalgulecha, a minor nit. I will take a pass in some time.

apb7 commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+describe('Player position service', function() {
+  beforeEach(module('oppia'));
+
+  var pts, pps, sof;
```
Incorrect variable declared, `sof`, should be `scof`.

ujjwalgulecha replied:
changed

apb7 commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+    expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(true);
+  });
+
+  it('should make learner submitted as false when navigation button is clicked',
```
`...should record answer submission by the learner as false...`

ujjwalgulecha replied:
changed

apb7 commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+      expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(false);
+    });
+
+  it('should set display index card to given value', function() {
```
`...displayed card index...`

ujjwalgulecha replied:
changed

apb7 commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+    expect(pps.getCurrentStateName()).toBe('First state');
+  });
+
+  it('should not change index if its same as previous', function() {
```
Which index? Please use clear descriptive language in tests. It should be: `...not change displayed card index if it is the same as the previously displayed card index `
Also, just curious -- what does this test bring to the coverage? Can you please check?

ujjwalgulecha replied:
it does not add to the coverage but tests the logic of the code.

ujjwalgulecha replied:
changed

apb7 reviewed (changes requested):
@ujjwalgulecha, ptal, thanks!

bansalnitish commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+  }));
+
+  it('should record answer submission correctly', function() {
+    pps.recordAnswerSubmission();
```
Ensure that the value is false before recording an answer submission.
```
    expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(false);
    pps.recordAnswerSubmission();
    expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(true);
```

ujjwalgulecha replied:
changed

bansalnitish commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+  it('should make learner submitted as false when navigation button is clicked',
+    function() {
+      pps.recordNavigationButtonClick();
+      expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(false);
```
This value is by default false. So, to make the test better, set the value to true by recording an answer and then check if it is set to false on navigation button click.

```
      expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(false);
      pps.recordAnswerSubmission();
      expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(true);
      pps.recordNavigationButtonClick();
      expect(pps.hasLearnerJustSubmittedAnAnswer()).toBe(false);
```

ujjwalgulecha replied:
changed

bansalnitish commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+    callBack = function() {};
+    pps.init(callBack);
+    pps.setDisplayedCardIndex(0);
+    expect(pps.getCurrentStateName()).toBe('First state');
```
Check this with multiple cards too i.e. add one more card and check if this works correctly.
```
    pts.addNewCard(scof.createNewCard(
      'Second state', 'Content HTML',
      '<oppia-text-input-html></oppia-text-input-html>'));
    pps.setDisplayedCardIndex(1);
    expect(pps.getCurrentStateName()).toBe('Second state');
```

ujjwalgulecha replied:
changed

bansalnitish commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+
+  it('should not change index if its same as previous', function() {
+    callBack = function() {};
+    pps.init(callBack);
```
Ensure that the index is null initially and also see if it is updated when you change it to a different value not same as the previous one.

```
    callBack = function() {};
    expect(pps.getDisplayedCardIndex()).toBe(null);
    pps.init(callBack);
    pps.setDisplayedCardIndex(4);
    pps.setDisplayedCardIndex(4);
    expect(pps.getDisplayedCardIndex()).toBe(4);
    pps.setDisplayedCardIndex(5);
    expect(pps.getDisplayedCardIndex()).toBe(5);
```

ujjwalgulecha replied:
the last 2 lines are not required in my opinion as they are being tested in another unit test(`should set display index card to given value`). What do you think?

ujjwalgulecha replied:
changed

bansalnitish replied:
I feel it is fine to have the last two lines here as well to ensure that if a new index is given after a repeated index, it changes.

ujjwalgulecha replied:
sure, changed it

bansalnitish commented on core/templates/dev/head/pages/exploration_player/PlayerPositionServiceSpec.js:
```diff
@@ -0,0 +1,64 @@
+  it('should set display index card to given value', function() {
+    callBack = function() {};
+    pps.init(callBack);
+    pps.setDisplayedCardIndex(4);
```
Ensure that the index is null initially.
```
    callBack = function() {};
    expect(pps.getDisplayedCardIndex()).toBe(null);
    pps.init(callBack);
    pps.setDisplayedCardIndex(4);
    expect(pps.getDisplayedCardIndex()).toBe(4);
```

ujjwalgulecha replied:
changed

bansalnitish reviewed (changes requested):
Hi @ujjwalgulecha, Thanks for the PR. I have left a review, PTAL. Thanks!

ujjwalgulecha committed:
Merge branch 'develop' of github.com:oppia/oppia into test_frontend_services

ujjwalgulecha committed:
address comments

ujjwalgulecha committed:
minor refactor

ujjwalgulecha commented:
@ankita240796 @bansalnitish @apb7 I have addressed all the comments. Please have a look and let me know if there's anything else that needs to be changed! Thanks

ujjwalgulecha committed:
Merge branch 'develop' of github.com:oppia/oppia into test_frontend_services

ujjwalgulecha committed:
minor refactor

ankita240796 reviewed (approved):
LGTM, Thanks @ujjwalgulecha!

bansalnitish reviewed (approved):
LGTM, Thanks @ujjwalgulecha!

bansalnitish commented:
Hi @apb7, this needs your approval, PTAL. Thanks!

bansalnitish commented:
Thanks @ujjwalgulecha, great work :)
